nvaccess/nvda
# Protect functions from using terminated OneCore instances

### Link to issue number:
N/A

### Summary of the issue:
NVDA can crash when resetting the configuration to factory defaults.
This is because NVDA terminates OneCore when resetting the config.
However, a race condition exists with the async function `speak`, where the python callback uses a terminated OneCore instance.

Steps to reproduce:
1. Run NVDA
1. In the Keyboard section of the Preferences dialog, turn "Speak command keys" on.
1. Reset your configuration to factory defaults
1. Note that NVDA crashes

### Description of how this pull request fixes the issue:
Protects usages of the OneCore instance by confirming the instance has not been terminated.
Wait until speech callbacks are completed before re-initializing.

### Testing strategy:


Manual testing

1. Run NVDA
1. In the Keyboard section of the Preferences dialog, turn "Speak command keys" on.
1. Reset your configuration to factory defaults
1. Note that NVDA does not crash.

### Known issues with pull request:
Other, somewhat related bugs, are logged as Errors during this process. See: #13580 

### Change log entries:
Bug fixes
```t2t
- NVDA no longer crashes when resetting the NVDA configuration to factory defaults while speak command keys is on. (#13569)
```

### Code Review Checklist:



  - description is up to date
  - change log entries
  - Unit tests
  - System (end to end) tests
  - Manual testing
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
